presto-main/src/main/java/com/facebook/presto/execution/QueryQueueRuleFactory.java

```diff
@@ -81,25 +82,38 @@ public QueryQueueRuleFactory(QueryManagerConfig config, ObjectMapper mapper)
                 rules.add(QueryQueueRule.createRule(rule.getUserRegex(), rule.getSourceRegex(), rule.getSessionPropertyRegexes(), rule.getQueues(), definitions));
             }
         }
-        checkIsDAG(rules.build());
+        checkIsTree(rules.build());
         this.selectors = rules.build();
     }
 
-    private static void checkIsDAG(List<QueryQueueRule> rules)
+    private static void checkIsTree(List<QueryQueueRule> rules)
     {
-        DirectedPseudograph<String, DefaultEdge> graph = new DirectedPseudograph<>(DefaultEdge.class);
+        DirectedGraph<String, DefaultEdge> graph = new DefaultDirectedGraph<>(DefaultEdge.class);
         for (QueryQueueRule rule : rules) {
             String lastQueueName = null;
             for (QueryQueueDefinition queue : rule.getQueues()) {
                 String currentQueueName = queue.getTemplate();
                 graph.addVertex(currentQueueName);
                 if (lastQueueName != null) {
+                    if (lastQueueName.equals(currentQueueName)) {
+                        throw new IllegalArgumentException("Queues must form a tree. Rules contain a loop for " + currentQueueName);
```

Comment:
@haozhun, actually I do need this check. I changed the graph from a pseudograph to a graph, otherwise the out-degree check below will fail when you have duplicate edges. However, directed graphs also disallow self edges